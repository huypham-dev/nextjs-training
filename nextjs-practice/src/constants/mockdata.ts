export const CATEGORIES = [
  {
    id: "1",
    name: "mobile",
    image:
      "https://images.prismic.io/nextjsexample/7853f863-6a12-4312-a6da-f930437f6d62_mobile.png",
  },
  {
    id: "2",
    name: "internet",
    image:
      "https://images.prismic.io/nextjsexample/cb71ce70-bb03-442b-99e8-a318255b81f3_3579187.png",
  },
];

export const ARTICLES = [
  {
    id: "1",
    createdAt: "2021-03-01T12:59:16.855Z",
    title: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
    author: "author 1",
    image:
      "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
    category: "internet",
    content:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
    slug: "slug 1",
    view: 90,
  },
  {
    id: "2",
    createdAt: "2021-03-01T12:59:16.855Z",
    title: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
    author: "author 1",
    image:
      "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
    category: "internet",
    content:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
    slug: "slug 1",
    view: 90,
  },
];
