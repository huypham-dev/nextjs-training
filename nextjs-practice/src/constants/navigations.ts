export const navigations = [
  { label: "Home", path: "/" },
  { label: "Blog", path: "/blog" },
  { label: "Contact", path: "/contact" },
];
