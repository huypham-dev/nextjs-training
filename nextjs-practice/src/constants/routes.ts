export const ROUTES = {
  home: {
    path: "/",
  },
  blog: {
    path: "/blog",
    withId: (id: string) => `/blog/${id}`,
  },
  post: {
    path: "/post",
    withId: (id: string) => `/post/${id}`,
  },
  contact: {
    path: "/",
  },
};
