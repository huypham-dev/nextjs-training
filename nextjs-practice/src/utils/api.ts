const BASE_URL = "https://5f7c244700bd74001690a4a7.mockapi.io";
export const COLLECTION = {
  CATEGORIES: "/categories",
  POSTS: "/posts",
};

export const API = {
  fetchAllCategories: async () => {
    const res = await fetch(`${BASE_URL}${COLLECTION.CATEGORIES}`);

    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },
  fetchPostsByCategory: async (categoryId) => {
    const res = await fetch(
      `${BASE_URL}${COLLECTION.POSTS}?category=${categoryId}`
    );

    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },
  fetchAllPost: async () => {
    const res = await fetch(`${BASE_URL}${COLLECTION.POSTS}`);
    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },
  fetchPostItem: async (id) => {
    const res = await fetch(`${BASE_URL}${COLLECTION.POSTS}/${id}`);
    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },

  fetchFeaturedPosts: async () => {
    const res = await fetch(
      `${BASE_URL}${COLLECTION.POSTS}?sortBy=view&order=desc&p=1&l=6`
    );
    return res.ok
      ? await res.json()
      : {
          errorCode: res.status,
          message: res.statusText,
        };
  },
};
