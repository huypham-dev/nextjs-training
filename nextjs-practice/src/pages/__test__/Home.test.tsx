import { shallow } from "enzyme";
import Home, { getStaticProps } from "..";
import { mockArticles, mockCategories } from "constants/mockdata";
import { API } from "utils/api";

const fetchAllCategories = jest.spyOn(API, "fetchAllCategories");
const fetchFeaturedPosts = jest.spyOn(API, "fetchFeaturedPosts");

describe("Home page", () => {
  it("Snapshot", () => {
    const container = shallow(<Home categoryList={[]} featuredPosts={[]} />);
    expect(container).toMatchSnapshot();
  });

  it("getStaticProps should call api", async () => {
    fetchAllCategories.mockImplementationOnce(() =>
      Promise.resolve(mockCategories)
    );
    fetchFeaturedPosts.mockImplementationOnce(() =>
      Promise.resolve(mockArticles)
    );
    const response = await getStaticProps({});

    expect(response).toEqual(
      expect.objectContaining({
        props: {
          categoryList: mockCategories,
          featuredPosts: mockArticles,
        },
      })
    );
  });

  it("getStaticProps should handle error", async () => {
    fetchAllCategories.mockImplementationOnce(() =>
      Promise.resolve({
        errorCode: 404,
        message: "Not found",
      })
    );
    fetchFeaturedPosts.mockImplementationOnce(() =>
      Promise.resolve({
        errorCode: 404,
        message: "Not found",
      })
    );
    const response = await getStaticProps({});

    expect(response).toEqual(
      expect.objectContaining({
        props: {
          categoryList: [],
          featuredPosts: [],
        },
      })
    );
  });
});
