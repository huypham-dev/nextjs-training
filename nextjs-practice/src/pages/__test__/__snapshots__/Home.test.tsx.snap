// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page Snapshot 1`] = `
<Fragment>
  <Head>
    <title>
      Home
    </title>
  </Head>
  <Memo(Introduce) />
  <Memo(Featured)
    featuredPosts={Array []}
  />
  <Memo(Categories)
    data={Array []}
  />
</Fragment>
`;
